### DISCUSSION TITLE
LangSmith along wih Llamaindex

### DISCUSSION THREAD

**Discussion by mansats** [run-llama/llama_index#12327] [General]
I have found the capabilites of LangSmith are interesting, but Can I use LangSmith to trace the RAG app built completely using LlamaIndex? 

I couldn't see LangSmith in the tracing section https://docs.llamaindex.ai/en/stable/module_guides/observability/

Please suggest.

**Comment by mansats** [comment]
Hi, Appreciate your feedback

**Comment by tituslhy** [comment]
Hello! 

Noticed nobody has replied here so I'll just quickly add an example on how to do this. This example requires an OpenAI API key and a LangSmith API Key. Hope this helps.

```
import os
from dotenv import load_dotenv, find_dotenv
from llama_index.core import VectorStoreIndex, SimpleDirectoryReader, Settings
from llama_index.llms.openai import OpenAI
from llama_index.embeddings.openai import OpenAIEmbedding
from langsmith import traceable

_ = load_dotenv(find_dotenv())

## Setup your RAG query engine
Settings.llm = OpenAI(model="...")
Settings.embed_model = OpenAIEmbedding(model="...")
documents = SimpleDirectoryReader("your folder here").load_data()
index = VectorStoreIndex.from_documents(documents)
query_engine = index.as_query_engine()

@traceable(type="tool", name="LlamaIndex RAG")
def ask_query_engine(query: str):
       return query_engine.query(query)
```

You'll now be able to see the trace on LangSmith along with the nodes retrieved from the vector store that the LlamaIndex query engine uses as its context. Adding a screenshot below:

<img width="1155" height="950" alt="image" src="https://github.com/user-attachments/assets/afde96e8-0775-4c48-ad40-fe4ef8bfca38" />
